src/main/java/uk/gov/ch/developer/docs/controller/developer/UserCallbackController.java

```diff
@@ -41,30 +57,63 @@ public String getCallback(@RequestParam("state") String state,
             // Nonce values
         }
 
-        WebClient webClient = WebClient.create();
+        final Session chSession = (Session) httpServletRequest
+                .getAttribute(SessionHandler.CHS_SESSION_REQUEST_ATT_KEY);
 
-        webClient.post().uri(URI.create(identityProvider.getTokenUrl()))
-                .contentType(MediaType.APPLICATION_FORM_URLENCODED)
-                .accept(MediaType.APPLICATION_JSON)
-                .body(BodyInserters.fromPublisher(Mono.just(getPostRequestBody(code)),
-                        String.class))
-                .retrieve().bodyToMono(OAuthToken.class)
-                .subscribe(s -> LOGGER.trace("Access Token: " + s.getAccessToken()));// TODO remove
-                                                                                     // this log
-                                                                                     // line
+        Map<String, Object> signInData = new HashMap<>();
+
+        LOGGER.debug("Getting User Profile....");
+
+        OAuthToken oauthToken = getOAuthToken(code);
+        UserProfileResponse userProfileResponse = getUserProfile(oauthToken);
+
+        if (userProfileResponse != null) {
+
+            LOGGER.debug("Processing Response");
+            oauthToken.setAccessToken(signInData);
+            userProfileResponse.setUserProfile(signInData);
+            
+            signInData.put(SessionKeys.SIGNED_IN.getKey(), 1);
+            chSession.getData().put(SessionKeys.SIGN_IN_INFO.getKey(), signInData);
+
+        }
+
+        LOGGER.debug("Redirecting...");
 
         return ("redirect:" + identityProvider.getRedirectUriPage());// TODO redirect back to page
                                                                      // where sign-in was initiated
     }
 
-    private String getPostRequestBody(String code) {
-        final StringBuilder sb = new StringBuilder();
-        sb.append("code=").append(code).append("&client_id=").append(identityProvider.getClientId())
-                .append("&client_secret=").append(identityProvider.getClientSecret())
-                .append("&redirect_uri=").append(identityProvider.getRedirectUri())
-                .append("&grant_type=authorization_code");
+    private UserProfileResponse getUserProfile(OAuthToken tokenResponse) {
+        LOGGER.debug("Requesting User Profile");
+        
+        final URI profileUrl = URI.create(identityProvider.getProfileUrl());
+        
+        return webClient.get()
+                .uri(profileUrl)
```

Comment:
It might be better to abstract this business logic, into the OAuth2 class, as it really should know what is going on. It also means that if the logic is changed in the OAuth processes, we don't need to change controllers.
Compare with the SignIn controller for preparation and retrieving of data.